File: chaeyun/3_palindrome/linked-list-palindrome/linked-list-palindrome.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head==None or head.next==None: 
            return head #If the list is empty or there is only one element then return it as it is
        prev=head
        curr=prev.next
        prev.next=None #After reversal first element will point to None
        temp=curr.next
        while curr.next:
            curr.next=prev
            prev=curr
            curr=temp
            temp=curr.next
        curr.next=prev #To connect last element to second last element
        return curr
    
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow=fast=head
        while fast.next and fast.next.next:
            slow=slow.next
            fast=fast.next.next
        temp1=self.reverseList(slow)
        temp2=head
        while temp1 and temp2:
            if temp1.val!=temp2.val:
                return False
            temp1=temp1.next
            temp2=temp2.next
        return True

File: chaeyun/3_palindrome/linked-list-palindrome/linked-list-palindrome.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
#include <stdio.h>
typedef int bool;
#define True  1
#define False 0

struct ListNode {
    int val;
    struct ListNode *next;
};

bool isPalindrome(struct ListNode* head){
        if(!head)   return False;
        if(!head->next) return True;

        struct ListNode *rev_head = NULL , *ptr = head;
        while(ptr != NULL)
        {
            struct ListNode *temp = (struct ListNode*)malloc(sizeof(struct ListNode));
            temp->val = ptr->val;
            temp->next = rev_head;
            rev_head = temp;
            ptr = ptr->next;
        }

        while(head && rev_head)
        {
            if(head->val != rev_head->val)  return False;
            else
            {
                head = head->next;
                rev_head = rev_head->next;
            }
        }
        return True;

}